File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
class Solution:

    def findK(self, start:ListNode, k:int):
        while start != None and k > 1:
            start = start.next
            k -= 1
        return start

    def reverse(self, start:ListNode, end: ListNode):
        prev = None
        cur = start
        while cur != end:
            temp = cur.next 
            cur.next = prev
            prev = cur 
            cur = temp
        end.next = prev
        return start


    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head == None:
            return None

        wait_pointer = ListNode(-1, None) 
        first_head = ListNode(-1, self.findK(head, k))

        while True:
            k_node = self.findK(head, k)
            if k_node == None:
                wait_pointer.next = head
                break
            wait_pointer.next = k_node
            next_group = k_node.next
            wait_pointer = self.reverse(head, k_node)
            head = next_group

        return first_head.next

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
struct resultPair {
    ListNode* newhead;
    ListNode* newtail;
    ListNode* nexthead;
    void print(){
        bool didprint = false;
        if(newhead != nullptr){
            cout<<newhead->val<<" ";
            didprint = true;
        }
        if(newtail != nullptr){
            cout<<newtail->val<<" ";
            didprint = true;
        }
        if(nexthead != nullptr){
            cout<<nexthead->val<<" ";
            didprint = true;
        }
        if(didprint) cout<<endl;
    }
};

class Solution {
public:

    resultPair reverse(ListNode* head, int k){
        ListNode anchor = ListNode();
        anchor.next = head;
        int counter = 1;
        while(head != nullptr && counter < k){
            counter += 1;
            head = head->next;
        }
        if(counter < k || head == nullptr){
            return {anchor.next, head, nullptr};
        }
        head = anchor.next;
        ListNode* prev = nullptr;
        counter = 1;

        while(counter <= k){
            counter += 1;
            ListNode* temp = head->next;
            head->next = prev;
            prev = head;
            head = temp;
        }
        return {prev, anchor.next, head};
    }

    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode anchor = ListNode(-1, head);
        ListNode* prev = &anchor;

        while(prev != nullptr && head != nullptr) {
            resultPair rp = reverse(head, k);
            prev->next = rp.newhead;
            prev = rp.newtail;
            head = rp.nexthead;
        }
        return anchor.next;
    }
};

/**
 * 1,2,3
 * k = 2
 * 
 */